lib/FrontendTool/FrontendTool.cpp

```diff
@@ -262,6 +263,185 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+    ModuleDecl::ImportFilter allImports = ModuleDecl::ImportFilterKind::Public;
+    allImports |= ModuleDecl::ImportFilterKind::Private;
+    allImports |= ModuleDecl::ImportFilterKind::ImplementationOnly;
+    allImports |= ModuleDecl::ImportFilterKind::SPIAccessControl;
+    allImports |= ModuleDecl::ImportFilterKind::ShadowedBySeparateOverlay;
+    filter = {allImports};
+  }
+
+  SmallVector<ModuleDecl::ImportedModule, 8> importList;
+  module->getImportedModules(importList, filter.getValue());
+
+  for (ModuleDecl::ImportedModule &import : importList)
+    imports.insert(import.importedModule);
+}
+
+/// Helper type for computing (approximate) information about ABI-dependencies.
+///
+/// This misses out on details such as typealiases and more.
+/// See the "isImportedDirectly" field above for more details.
+class ABIDependencyEvaluator {
+  /// Map of ABIs exported by a particular module, excluding itself.
+  ///
+  /// For example, consider (primed letters represent Clang modules): 
+  /// ```
+  /// - A is @_exported-imported by B
+  /// - B is #imported by C' (via a compiler-generated umbrella header)
+  /// - C' is @_exported-imported by C (Swift overlay)
+  /// - D' is #imported by E'
+  /// - D' is @_exported-imported by D (Swift overlay)
+  /// - E' is @_exported-imported by E (Swift overlay)
+  /// ```
+  ///
+  /// Then the \c abiExportMap will be
+  /// ```
+  /// { A: {}, B: {A}, C : {B}, C': {B}, D': {}, D: {}, E: {D'}, E': {D'} }
+  /// ```
+  /// TODO: Not 100% sure about this ^. Need to fix before merging PR.
+  ///
+  /// WARNING: Use `addToABIExportMap` instead of inserting directly.
+  llvm::DenseMap<ModuleDecl *, llvm::DenseSet<ModuleDecl *>> abiExportMap;
+  
+  /// Stack for depth-first traversal.
+  SmallVector<ModuleDecl *, 32> searchStack;
+
+  llvm::DenseSet<ModuleDecl *> visited;
+
+  /// Correct way to add entries to the abiExportMap.
+  /// 
+  /// The `getTopLevelModule()` is needed to avoid exploding `abiExportMap` with
+  /// contents from submodules. Since we only care about framework-level
+  /// dependencies, this is ok.
+  void addToABIExportMap(ModuleDecl *module, ModuleDecl *reexport) {
+    if (reexport->isClangModule())
+      abiExportMap[module].insert(reexport->getTopLevelModule());
+    else
+      abiExportMap[module].insert(reexport);
+  }
+  
+  /// Mark `importedModule` as exposed for `module`'s ABI.
+  void reexposeImportedABI(ModuleDecl *module, ModuleDecl *importedModule) {
+    addToABIExportMap(module, importedModule);
+    for (auto reexportedModule: abiExportMap[importedModule])
+      addToABIExportMap(module, reexportedModule);
+  }
+
+  /// Recursive step in computing ABI dependencies.
+  ///
+  /// Use this method instead of using the `forClangModule`/`forSwiftModule`
+  /// methods.
+  bool computeABIDependenciesForModule(ModuleDecl *module) {
+    for (auto soFar: searchStack) {
+      if (soFar == module) {
+        // Cycle in import graph, aaaaa, bail!
+        return false;
```

Comment:
1. I don't think asserting here is a good idea. We used to have lots of asserts in the module trace emission code previously and we slowly replaced them with `llvm::report_fatal_error` and then no-ops one-by-one due to weird difficult-to-reproduce edge cases happening in the build. It causes a bunch of trouble that needs to be fixed quickly. Since this is a supplementary output, it's better to not crash (and fail the build, blocking a lot of people) and instead provide whatever output we can on a best-effort basis. You will see similar notes in `computeSwiftModuleTraceInfo` (already existing, not part of this patch).
2. I don't think it can happen, but given my track record here, I think I'm probably wrong.

I should write down a longer justification in the comment here.